File: LeetCode/magnetic-force-between-two-balls.java
class Solution {
    public int maxDistance(int[] position, int m) {
        Arrays.sort(position);
        int l=1;
        int h=position[position.length-1]-position[0];
        while (l<=h){
            int mid=l+(h-l)/2;
            if(solve(mid,position,m)){
                l=mid+1;
            }
            else{
                h=mid-1;
            }
        }
        return h+1;
    }
    public boolean solve(int distance,int[] position,int m){
        int c=1;
        int last = position[0];
        for(int i=1;i<position.length;i++){
            if(position[i]-last >distance){
                c++;
                last = position[i];
                if(c==m){
                    return true;
                }
            }
        }
        return false;
    }
}

File: LeetCode/magnetic-force-between-two-balls.py
import math
class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        def solve(dist):
            c=1
            last=position[0]
            for i in position[1:]:
                if i-last>=dist:
                    c+=1
                    last=i
                    if c==m:
                        return True
            return False
        position.sort()
        l=1
        h=position[-1]-position[0]
        while l<=h:
            mid=l+(h-l)//2
            if solve(mid):
                l=mid+1
            else:
                h=mid-1
        return h